### PR TITLE
Issue78

### PR DESCRIPTION
**Pull request by maektwain**

### CODE REVIEW

**Comment by cryptomental** [comment]
Hi @maektwain Travis checks failed:

{code}
Summary of all failing tests
 FAIL  test/util/validations.test.js (50.296s)
  ● CheckContractAction › should validate contract based on the valid address provided
    AssertionError: expected {} to equal ''
      
      at Object.it (test/util/validations.test.js:16:42)
          at new Promise (<anonymous>)
      at Promise.resolve.then.el (node_modules/p-map/index.js:46:16)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:118:7)
  ● CheckContractAction › should invalidate the  contract based on the invalid address provided
    AssertionError: expected {} to equal 'Not a valid smart contract address'
      
      at Object.it (test/util/validations.test.js:26:42)
          at new Promise (<anonymous>)
      at Promise.resolve.then.el (node_modules/p-map/index.js:46:16)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:118:7)
Test Suites: 1 failed, 26 passed, 27 total
Tests:       2 failed, 110 passed, 112 total
{code}

This needs to be fixed before the pull request can be merged.

**Comment by maektwain** [comment]
The problem with the above test , I need to fake Geth , or if not fake geth then create a fake store which can get the geth. Both are open problems.

**Comment by cryptomental** [comment]
@maektwain, please:

* Open a ticket for 'open problem A' , describe it and suggest an approach to solve it if you can think of any solution.
* Open a ticket for 'open problem B' , describe it and suggest an approach to solve it if you can think of any solution.

A GitCoin issue will be opened to the public to help with those two.

Also 'now working' statement is not enough. Please provide source code and debug logs that lead to the failure. You may also want to attach a few screenshots if needed.

**Comment by pelsasser** [comment]
@maektwain please see @cryptomental comments and respond to the status here so we can move this along

**Comment by maektwain** [comment]
I would like to give it a final try today, else I will post the issue.

**Comment by perfectmak** [comment]
Well done @maektwain, I really believe we shouldn't have to mock geth especially on unit tests.
And this issue is kinda a pointer that we need to modularize more and also adequately pass in all dependencies (especially side effect(ing) dependencies) and not accessing it globally.

So here is a suggestion on how this might be fixed. Inside the `checkContract` validation, what you are after is the `web3Instance` inside the store. So instead of 
```js
const web3 = store.getState().web3.web3Instance;
```
The `checkContract` function should be modified such that the web3Instance will be passed in as a dependency, so perhaps something like this:
```js
const checkContract = (web3, rule, value, callback) {
  // now you have access to web3 and 
 // this side effecting dependency can be mocked or altered from the calling code
}
```
Next inside the `DeployContractField` where this is used, we can also have it get access to the `web3` dependency as a property. So whatever HOC that will be using a contract field that required web3 would also need to have access to the web3 dependency.

Basically, we should avoid accessing the store directly from inside these components. Components should be dumb and their state should be driven solely by their state and properties. So store values required by components should be passed to them on dispatch as properties.

This way components would be easily testable.

**Comment by maektwain** [comment]
@pelsasser @eswarasai I have just found a quick hack to work out with the test provided in the validations.

**Comment by maektwain** [comment]
@cryptomental Is this the right way?

**Comment by cryptomental** [comment]
Yes, @maektwain I agree with @perfectmak , can be merged if accepted by @pelsasser .

File: src/util/validations.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+let web3;
+
+if (typeof web3 !== 'undefined') {
+  web3 = new Web3(web3.currentProvider);
+} else {
+  // set the provider you want from Web3.providers
+  web3 = new Web3(new Web3.providers.HttpProvider("http://localhost:9545"));
+}
+
+async function checkContract  (rule,value,callback){
```
Comment (eswarasai) [inline src/util/validations.js]: @maektwain -- So the idea is to pass the `web3` reference from wherever this function is being called. I don't think there's any need for the `web3` stuff in the `validations.js`. It should purely contain the validation functions. And also `checkContract()` function doesn't need `rule` to be passed as param as you aren't using it in the underlying method.

Comment (maektwain) [reply src/util/validations.js]: Yeah , rule can be modified and `const web3 = store.getState().web3.web3Instance` can be passed like this but the problem is that it requires Dom to be created while in the test.

File: src/util/validations.js [hunk 2]
```diff
@@ -1,34 +1,37 @@
-
-import  Web3 from 'web3';
+/*import Web3 from 'web3';
 
 let web3;
```
Comment (perfectmak) [inline src/util/validations.js]: If `web3` is not been passed won to the `checkContract()` function, then we don't need this global web3 anymore. Right?

Comment (maektwain) [reply src/util/validations.js]: So we change the logic because in case if we are using Rinkeby netowrk  we might need to check against '0x'.

**Comment by pelsasser** [comment]
@maektwain - any updates to this? would love to be able to get it merged and you paid out.

**Comment by maektwain** [comment]
@eswarasai We can close this as I have created another PR #114
